### PR TITLE
[FunctionAttrs] Determine underlying object by `getUnderlyingObjectAggressive`

### LINKED ISSUE #119573: Missing CSE caused by for loop on slice
While examining a weird performance regression on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Odd.20Performance.20Phenomenon), we noticed a peculiar thing caused by a simple for loop.

Consider this code:
```rust
#[inline(never)]
fn add_index(bools: &[bool]) -> usize {
    let mut count: usize = 0;
    for i in 0..bools.len() {
        if bools[i] {
            count += 1;
        }
    }
    count
}

#[inline(never)]
fn add_iter(bools: &[bool]) -> usize {
    let mut count: usize = 0;
    for b in bools {
        if *b {
            count += 1;
        }
    }
    count
}
```
These two functions do the same thing, but the first one uses slice indexing, while the other one uses a more natural for loop. One would expect that the second version will optimize better (or same).

However, when these functions are called multiple times, only the first function is eligible for CSE (common subexpression elimination):
```rust
pub fn foo1(bools: &[bool]) {
    // CSE applied
    println!("a: {}", add_index(bools));
    println!("b: {}", add_index(bools));
}

pub fn foo2(bools: &[bool]) {
    // CSE not applied
    println!("a: {}", add_iter(bools));
    println!("b: {}", add_iter(bools));
}
```
Link to [Godbolt](https://godbolt.org/z/bcdf7hn7G).

When examining these two functions, I noticed that the indexed version uses `nocapture` for its argument, while the for loop version doesn't:
```
define noundef i64 @add_index(ptr noalias nocapture noundef nonnull readonly align 1 %bools.0, i64 noundef %bools.1) unnamed_addr #0 !dbg !7 {

define noundef i64 @add_iter(ptr noalias noundef nonnull readonly align 1 %bools.0, i64 noundef %bools.1) unnamed_addr #1 !dbg !57 {
```
which might be the reason of why CSE isn't applied.

It seems weird that the canonical way of iterating over a slice produces worse code (or at least function attributes) than manual indexing.

### PR DESCRIPTION
**Pull request by dianqk**
Thanks to #99509, we can fix https://github.com/rust-lang/rust/issues/119573 too.

### CODE REVIEW

File: llvm/lib/Transforms/IPO/FunctionAttrs.cpp [hunk 1]
```diff
@@ -118,7 +118,7 @@ static void addLocAccess(MemoryEffects &ME, const MemoryLocation &Loc,
   if (isNoModRef(MR))
     return;
 
-  const Value *UO = getUnderlyingObject(Loc.Ptr);
+  const Value *UO = getUnderlyingObjectAggressive(Loc.Ptr);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/IPO/FunctionAttrs.cpp:121]: Compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=c1b70fa5bfea973d4141e27cf9668e9325609e19&to=9c73bbf882a265931e4669e6da2a1c99acbb4efe&stat=instructions:u

**Review by dtcxzyw — APPROVED** [review]
LGTM if the original rust issue can be fixed :)

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/IPO/FunctionAttrs.cpp [hunk 2]
```diff
@@ -118,7 +118,7 @@ static void addLocAccess(MemoryEffects &ME, const MemoryLocation &Loc,
...
   assert(!isa<AllocaInst>(UO) &&
          "Should have been handled by getModRefInfoMask()");
```
Comment (nikic) [inline llvm/lib/Transforms/IPO/FunctionAttrs.cpp:123]: Should convert this into `if(isa<AllocaInst>(UO)) return`.

Comment (DianQK) [reply llvm/lib/Transforms/IPO/FunctionAttrs.cpp:123]: Thanks! I see it in the logs now. It doesn't seem to require additional test cases, so I will submit it directly later.

Comment (DianQK) [reply llvm/lib/Transforms/IPO/FunctionAttrs.cpp:123]: Reland with a test case: https://github.com/llvm/llvm-project/commit/559be8e2b55afbf6736fed646542fb25793f0f7e.

**Pull request merged**